proxy/src/transport/tls/config.rs

```diff
@@ -367,3 +342,886 @@ fn set_common_settings(versions: &mut Vec<rustls::ProtocolVersion>) {
+        end_entity_cert.write_all(b"I am not real end entity cert")
+            .expect("write to end entity cert");
+        end_entity_cert.sync_all().expect("sync end entity cert");
+
+        symlink_file(data_path.clone().join(PRIVATE_KEY), cfg.private_key)
+            .expect("symlink private key");
+        symlink_file(data_path.clone().join(END_ENTITY_CERT), cfg.end_entity_cert)
+            .expect("symlink end entity cert");
+        symlink_file(data_path.clone().join(TRUST_ANCHORS), cfg.trust_anchors)
+            .expect("symlink trust anchors");
+
+        let (watch, bg) = watch_stream(stream);
+        rt.spawn(Box::new(bg));
+
+        trust_anchors.write_all(b"Trust me on this :)")
+            .expect("write to trust anchors again");
+        trust_anchors.sync_all()
+            .expect("sync trust anchors again");
+
+        let next = watch.into_future().map_err(|(e, _)| e);
+        let (item, watch) = rt.block_on_for(Duration::from_secs(5), next)
+            .expect("first change");
+        assert!(item.is_some());
+        println!("saw first change");
+
+        end_entity_cert.write_all(b"This is the end of the end entity cert :)")
+            .expect("write to end entity cert again");
+        end_entity_cert.sync_all()
+            .expect("sync end entity cert again");
+
+        let next = watch.into_future().map_err(|(e, _)| e);
+        let (item, watch) = rt.block_on_for(Duration::from_secs(5), next)
+            .expect("second change");
+        assert!(item.is_some());
+        println!("saw second change");
+
+        private_key.write_all(b"Keep me private :)")
+            .expect("write to private key");
+        private_key.sync_all()
+            .expect("sync private key again");
+
+        let next = watch.into_future().map_err(|(e, _)| e);
+        let (item, _) = rt.block_on_for(Duration::from_secs(5), next)
+            .expect("third change");
+        assert!(item.is_some());
+        println!("saw third change");
+    }
+
+    fn test_detects_double_symlink_retargeting(
+        fixture: Fixture,
+        stream: impl Stream<Item = (), Error=()> + 'static,
+    ) {
+        let Fixture { cfg, dir, mut rt } = fixture;
+
+        let real_data_path = dir.path().join("real_data");
+        let real_data_path_2 = dir.path().join("real_data_2");
+        let data_path = dir.path().join("data");
+        fs::create_dir(&real_data_path).expect("create data dir");
+        fs::create_dir(&real_data_path_2).expect("create data dir 2");
+        symlink_dir(&real_data_path, &data_path).expect("create data dir symlink");
+
+        let mut trust_anchors = File::create(real_data_path.clone().join(TRUST_ANCHORS))
+            .expect("create trust anchors");
+        println!("created {:?}", trust_anchors);
+        trust_anchors.write_all(b"I am not real trust anchors")
+            .expect("write to trust anchors");
+        trust_anchors.sync_all().expect("sync trust anchors");
```

Comment:
Is it really valuable to have the `expect()` messages everywhere? Why not just `unwrap()` and make the code easier to read? If we need to know what failed we can use the backtrace.

It seems like this `File::create...write_all...sync_all()` pattern should be encapsulated into a function that's reused.

Similarly, we should encapsulate the larger logic of "create a file in one directory and then create a symlink to it in another directory" and the larger logic "create the file and the first symlink and then create the second symlink."

More generally, I can't really carefully review this much repetitive testing code. I tried last night and again today and it's just overwhelming. So, the less code there actually is in the test, and the more reusable functions are reused, the better.

Reply:
> It seems like this `File::create...write_all...sync_all()` pattern should be encapsulated into a function that's reused.

I agree. I  intended to do this in a separate follow-up PR, as I wanted to get this up for review sooner, but I'd be happy to make that change now if you prefer.